### PR TITLE
rbac: add uri_template for path matching

### LINKED ISSUE #30724: rbac: url fragment matchers
*Title*: rbac: url fragment matchers

*Description*:

In HTTP APIs, it is extremely common to have paths of the form `/orders/{storeid}/{orderId}/view`.

It is useful to be able to safely match over these URLs in RBAC.

RBAC [URL matching](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/path.proto#envoy-v3-api-msg-type-matcher-v3-pathmatcher) is currently a `StringMatcher`.

One could attempt to match these urls with regex. For example, I believe the above is roughly `/orders/[^/]+/[^/]+/view`.

However, regex matching against URLs feels fraught with issues. URL parsing confusion attacks are extremely common, and has had a lot of consideration in Envoy ([path normalization](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-normalize-path), etc).

It would be great to have a first class way to match these in regex. I could see two approaches for this:

1. Adopt a new URL-aware matching setup. https://github.com/cncf/xds/pull/40/files may provide some previous discussion
2. Document, test, and support a canonical regex representation for URLs of this nature. As part of this, it would be desired that the security of this regex usage is treated the same as RBAC more broadly (e.g if the canonical regex has issues, its treated as a high severity CVE).

### PR DESCRIPTION
**Pull request by kozjan**
Commit Message: rbac: add uri_template for path matching 
Additional Description: Added `uri_template`  with  `envoy.path.match` extension category to allow matching with URI templates in RBAC.
Risk Level: low
Testing: unit, integration
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Fixes #30724

### CODE REVIEW

File: api/envoy/config/rbac/v3/rbac.proto
Comment (htuch) [inline api/envoy/config/rbac/v3/rbac.proto:366]: @envoyproxy/api-shepherds I'm wondering if type.matcher.v3.PathMatcher should absorb envoy.extensions.path.match.uri_template.v3.UriTemplateMatchConfig somehow?

The problem is this is core depending on some extension. But, maybe we want to consider promoting URI template matching to core? Not sure, but when I see a change like this it seems worth asking.

Comment (htuch) [reply api/envoy/config/rbac/v3/rbac.proto:366]: I.e. allow URI template matcher to be plugged in. In any case, we should not be referencing the extension config directly here and instead should have something like https://github.com/envoyproxy/envoy/blob/b039d8a316a081d1eae3e3222e045b749d26765e/api/envoy/config/route/v3/route_components.proto#L601 if we don't end up putting it inside PathMatcher.

Comment (kozjan) [reply api/envoy/config/rbac/v3/rbac.proto:366]: hey, I'm going on holiday so I will rework it once I'm back, probably next year

File: api/envoy/config/rbac/v3/rbac.proto
Comment (sschepens) [inline api/envoy/config/rbac/v3/rbac.proto:272]: we already have typed matchers, seems like adding another typedextensionconfig y kind of redundant? can't we simply create a matcher for `uri_template`?

File: api/envoy/config/rbac/v3/rbac.proto
Comment (sschepens) [inline api/envoy/config/rbac/v3/rbac.proto:276]: I don't think i like naming this `glob_path` given that it possible covers more features than just globbing, that's why it's actually called `uri_template` 🤔

File: api/envoy/config/rbac/v3/rbac.proto
Comment (kozjan) [reply api/envoy/config/rbac/v3/rbac.proto:272]: I don't understand the extension system well - does it mean that I would have to copy [uri_template_match.proto](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/path/match/uri_template/v3/uri_template_match.proto) and paste it next to [upstream_ip_port_matcher.proto](https://github.com/envoyproxy/envoy/tree/main/api/envoy/extensions/rbac/matchers/upstream_ip_port/v3) while adjusting the `// [#extension: envoy.path.match.uri_template.uri_template_matcher]` to also begin with `envoy.rbac.matchers` as in `upstream_ip_port_matcher.proto`?

File: api/envoy/config/rbac/v3/rbac.proto
Comment (kozjan) [reply api/envoy/config/rbac/v3/rbac.proto:276]: I will replace `glob` everywhere with `uri_template`.

File: api/envoy/config/rbac/v3/rbac.proto
Comment (kozjan) [reply api/envoy/config/rbac/v3/rbac.proto]: i did as i described in the comment

**Comment by kozjan** [comment]
I tried to push with prepush hooks, but shellcheck doesn't work on Mac i guess:
```
ERROR: /private/var/tmp/_bazel_jan.kozlowski/a20ce50f1167d001f45ad8fee633e1f8/external/com_github_aignas_rules_shellcheck/BUILD.bazel:25:6: configurable attribute "actual" in @com_github_aignas_rules_shellcheck//:shellcheck doesn't match this configuration: binaries for your platform could not be found
```
so i fixed everything that was before the shellcheck, but i'm not sure if that's all

**Comment by kozjan** [comment]
don't really see why some checks failed, envoy/prechecks build result look fine yet it failed, and in envoy/windows do not see any specific errors, so running a retest
/retest

File: api/envoy/config/rbac/v3/rbac.proto
Comment (htuch) [reply api/envoy/config/rbac/v3/rbac.proto]: No, we should not copy+paste like that. Before rolling back, @sschepens are you suggesting we just used the generic matcher capability here and plug into that?

I don't think we need a static entry for URI template, I don't think we need to grow the static match capabilities here, rather we do it via generic match on URL (as I was proposing and was present in https://github.com/envoyproxy/envoy/pull/31447/commits/528c89773c456a2e4f397a17ff4edb2a39b7fb6f, or we make URI template matching a generic matcher extension as well.

Comment (sschepens) [reply api/envoy/config/rbac/v3/rbac.proto]: > No, we should not copy+paste like that. Before rolling back, @sschepens are you suggesting we just used the generic matcher capability here and plug into that?
> 
> I don't think we need a static entry for URI template, I don't think we need to grow the static match capabilities here, rather we do it via generic match on URL (as I was proposing and was present in [528c897](https://github.com/envoyproxy/envoy/commit/528c89773c456a2e4f397a17ff4edb2a39b7fb6f), or we make URI template matching a generic matcher extension as well.

In all honesty I don't really like extensions since they make for a ver verbose usage, more so when we can have many matchers. I'd really like for URI matching to become a first class and probably be included without needing `TypedExtensionConfig`, but  I don't think that's going to happen. So, for me, it made more sense to make URI template a generic matcher extension and avoid adding another member to PathMatcher making use of the standard matcher extension mechanism.

Comment (htuch) [reply api/envoy/config/rbac/v3/rbac.proto]: I'd suggest opening a discussions with @envoyproxy/api-shepherds on the tradeoffs between extensibility and verbosity in configuration - in general we err towards more pluggable extension points and assume for the most part machine generation of configuration. xDS is sometimes read/written by humans, but we do not have any goal today around conciseness, in particular when it conflicts with generic extension points.

In this case, URI matching is a plugin for an existing extension class, so it seem arbitrary to opt to hardcode it into the RBAC API rather than the extension class.

File: api/envoy/config/rbac/v3/rbac.proto
Comment (htuch) [inline api/envoy/config/rbac/v3/rbac.proto:275]:
```suggestion
    // [#extension-category: envoy.path.match]
```

File: bazel/repository_locations.bzl
Comment (htuch) [inline bazel/repository_locations.bzl:1251]: I don't get why this is needed?

File: source/extensions/extensions_build_config.bzl
Comment (htuch) [inline source/extensions/extensions_build_config.bzl:432]: There's no need to declare a new matcher here, we can use the existing URI template path matcher extension?

Comment (kozjan) [reply source/extensions/extensions_build_config.bzl:432]: If I understand correctly, this is necessary if I want to use a different factory for rbac, the one that was originally implemented with URI template path matcher returns [absl::StatusOr\<Router::PathMatcherSharedPtr>](https://github.com/envoyproxy/envoy/blob/main/source/extensions/path/match/uri_template/config.h#L19-L23), while the type needed in RBAC is [MatcherConstSharedPtr](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/common/rbac/matchers.cc#L15-L16)

would it be better to use the same factory, but then wrap the `PathMatcherSharedPtr` into `MatcherConstSharedPtr`?

File: bazel/repository_locations.bzl
Comment (kozjan) [reply bazel/repository_locations.bzl:1251]: I added it because of an [error message](https://github.com/envoyproxy/envoy/actions/runs/7612248977/job/20729480652#step:8:628) in Envoy/Prechecks :
```
Dependency validation failed, please check metadata in bazel/repository_locations.bzl
Extension envoy.rbac.matchers.uri_template depends on com_github_google_flatbuffers but com_github_google_flatbuffers does not list envoy.rbac.matchers.uri_template in its allowlist
```
However, it seems to me as if the checks are bugged and checkout older code, because when I run it locally I don't get this error know - I think `envoy.rbac.matchers.uri_template` value comes from [source/extensions/extensions_build_config.bzl](https://github.com/envoyproxy/envoy/pull/31447/files/e1acfe4a996cce7813264e69efed8424ce7366e7#diff-5294c859ad138c7624de53e0f9566c0b630f5a9522ff357d23b2f0b17d6e3991)

File: api/envoy/config/rbac/v3/rbac.proto
Comment (htuch) [inline api/envoy/config/rbac/v3/rbac.proto:197]: LGTM, can you add an entry to https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.30/v1.30.0? Thanks.

Comment (kozjan) [reply api/envoy/config/rbac/v3/rbac.proto:197]: done, also made some small changes to tests and unnecessary null check, since it works the same either way

**Comment by kozjan** [comment]
don't know what to do with Envoy/Windows: https://github.com/envoyproxy/envoy/actions/runs/7651824181/job/20850377753#step:8:21416
looks like some Azure problem to me, hoped rerunning would fix it but nothing changed

**Comment by phlax** [comment]
>  looks like some Azure problem to me, hoped rerunning would fix it but nothing changed

its a backend (RBE) issue - its known but infrequent generally, and not obvious why its happening

ive kicked the CI again

**Comment by phlax** [comment]
@kozjan i think its failing because your branch is far behind

can you merge `main` please

**Comment by phlax** [comment]
hmm - the more i think about it the more im wondering why that would matter - and it looks like you have merged main relatively recently - and also it was passing previously

a bit of a mystery tbh - this is now failing every time, no other ci runs are 

i think we may have some misbehaving backend auth server cc @adisuissa

**Review by htuch — DISMISSED** [review]
LGTM for when CI passes.

**Comment by phlax** [comment]
https://github.com/envoyproxy/envoy/issues/32044

**Comment by kozjan** [comment]
hey, it passed today - do I have to contact someone for the second review?

**Review by htuch — APPROVED** [review]

**Pull request merged**
